### PR TITLE
Fix the POV button in spectator mode

### PR DESCRIPTION
**Pull request by Khalmee**
Pressing the Middle Mouse Button while spectating switches POV from 3rd to 1st person, just like how it works in vanilla.

### CODE REVIEW

**Comment by x3Karma** [comment]
Looks good but I think it shouldn't need to be in a separate file, maybe modify `cl_gamestate.nut` would be fine.

File: Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut
Comment (uniboi) [inline Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut:4]: The button callback needs to be only registered when spectating

Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut:4]: This doesnt support controllers?

File: Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut:9]: I need to look into this properly, but isn't spec_mode already bound? If so, why doesn't it already work?

File: Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut:1]: This needs a formatting pass

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]

**Comment by ASpoonPlaysGames** [comment]
> Looks good but I think it shouldn't need to be in a separate file, maybe modify `cl_gamestate.nut` would be fine.

I think ideally we put it in the same file as the spectator UI?  I don't see a reason why this needs to be specifically on the CLIENT vm

File: Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut
Comment (Khalmee) [reply Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut:4]: Never used a controller for more than 15 minutes so I don't really know what the input for controller should be

Comment (Khalmee) [reply Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut:4]: Registering and deregistering it would be pain, especially because all of the possible combinations of spectating/respawning/joining a game/switching players etc., so unless I manage do find proper callbacks for that I'd leave it as it is, since it only does something when spectating

File: Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut
Comment (ASpoonPlaysGames) [reply Northstar.Client/mod/scripts/vscripts/spectator_pov_button_fix.nut:9]: Just an update, it's handled in native and is a pain in the ass.
It works changing from first person to third person, but not the other way around?

This conversation could probably be marked as resolved

**Review by ASpoonPlaysGames — APPROVED** [review]
Works now in testing for me, and I did a formatting and code pass, so LGTM
(since I have commits on this PR now, another reviewer would be appreciated)

**Comment by DBmaoha** [comment]
This can be fixed by pure server-side, just add 0.1s spectator replay delay on third person spec mode.
in `mp/_spectator.gnut`, line 167
```
if ( target.IsPlayer() )
{		
	try
	{
		player.SetObserverTarget( target )
		player.StartObserverMode( OBS_MODE_CHASE )
		player.SetSpecReplayDelay( 0.1 ) // at least 0.1s specReplayDelay is required for client to realize player is spectating, and can send "spec_mode" command!
	}
	catch ( ex ) { }
}
```
the same file, in function ClientCommandCallback_spec_mode()
```
bool function ClientCommandCallback_spec_mode( entity player, array<string> args )
{
	// currently unsure how this actually gets called on client, works through console and has references in client.dll tho
	if ( player.GetObserverMode() == OBS_MODE_CHASE )
	{
		// set to first person spectate		
		player.SetSpecReplayDelay( FIRST_PERSON_SPECTATOR_DELAY )
		player.SetViewEntity( player.GetObserverTarget(), true )
		player.StartObserverMode( OBS_MODE_IN_EYE )
	}
	else if ( player.GetObserverMode() == OBS_MODE_IN_EYE )
	{	
		// set to third person spectate
		player.SetSpecReplayDelay( 0.1 ) // at least 0.1s specReplayDelay is required for client to realize player is spectating, and can send "spec_mode" command!
		player.StartObserverMode( OBS_MODE_CHASE )
	}
	
	return true
}
```

**Comment by F1F7Y** [comment]
> I found out that this can be fixed on pure server-side, by adding 0.1s spectator replay delay while in third person spec mode in `mp/_spectator.gnut`, line 167
> 
> ```js
> if ( target.IsPlayer() )
> {		
> 	try
> 	{
> 		player.SetObserverTarget( target )
> 		player.StartObserverMode( OBS_MODE_CHASE )
> 		player.SetSpecReplayDelay( 0.1 ) // at least 0.1s specReplayDelay is required for client to know that player is spectating, and they can send "spec_mode" client command
> 	}
> 	catch ( ex ) { }
> }
> ```
> 
> the same file, in function `ClientCommandCallback_spec_mode()`
> 
> ```js
> bool function ClientCommandCallback_spec_mode( entity player, array<string> args )
> {
> 	// currently unsure how this actually gets called on client, works through console and has references in client.dll tho
> 	if ( player.GetObserverMode() == OBS_MODE_CHASE )
> 	{
> 		// set to first person spectate		
> 		player.SetSpecReplayDelay( FIRST_PERSON_SPECTATOR_DELAY )
> 		player.SetViewEntity( player.GetObserverTarget(), true )
> 		player.StartObserverMode( OBS_MODE_IN_EYE )
> 	}
> 	else if ( player.GetObserverMode() == OBS_MODE_IN_EYE )
> 	{	
> 		// set to third person spectate
> 		player.StartObserverMode( OBS_MODE_CHASE )
> 		player.SetSpecReplayDelay( 0.1 ) // at least 0.1s specReplayDelay is required for client to know that player is spectating, and they can send "spec_mode" client command
> 	}
> 	
> 	return true
> }
> ```

Recon you could do a small Pr with these changes ?

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]
Would be cool if the above suggestions get implemented.

**Comment by ASpoonPlaysGames** [comment]
@Khalmee is this PR stale?

**Review by GeckoEidechse — CHANGES_REQUESTED** [review]
Superseded by #706

(will close once other one gets merged)

**Pull request closed without merging**
